### PR TITLE
Option to set 'non_blocking' for to(device) in BatchEncoding and BatchFeature

### PR DESCRIPTION
**Pull request by daniel-bogdoll**
Option to set 'non_blocking' for to(device) operation in BatchEncoding for performance improvements. Defaults to 'false', thus no behavioral changes.

# What does this PR do?

This minor PR adds the non_blocking option to the to() function.

Previous: def to(self, device: Union[str, "torch.device"]) -> "BatchEncoding":
New: def to(self, device: Union[str, "torch.device"], non_blocking: bool = False) -> "BatchEncoding":

Since non_blocking defaults to 'False', this PR does not introduce behavioral changes. 

I realized, when utilizing Zero Shot Object Detection models, that it was not possible to set this option, leading to sub-optimal performance during inference.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

- vision models: @amyeroberts, @qubvel

### CODE REVIEW

**Review by qubvel — COMMENTED** [review]
Hi @daniel-bogdoll, thanks for adding this! It looks great to me. Do you think it might be worth extending the same option to BatchFeature to ensure consistent capabilities?

**Comment by daniel-bogdoll** [comment]
Sure thing! Which tests would I need to run to make sure modifications in the to() function of BatchFeature get tested?

Just to make sure, I assume you refer to https://github.com/huggingface/transformers/blob/54be2d7ae87e873482b984cc956e165ca4dc0ba3/src/transformers/feature_extraction_utils.py#L206 ?

**Comment by qubvel** [comment]
Yes, I refer to this one, but not sure it's properly tested anywhere, I was able to find only `SequenceFeatureExtractionTestMixin`

**Comment by qubvel** [comment]
Maybe can do it as simple as 

```python
non_blocking = kwargs.get("non_blocking", False)
...
elif isinstance(v, torch.Tensor) and device is not None:
      new_data[k] = v.to(device=device, non_blocking=non_blocking)
...
```

**Comment by qubvel** [comment]
>Here device is derived from args rather than kwargs. Should this be extended in some way to also consider deriving non_blocking? Not sure where or how this is used.

I don't think so, maybe at some moment, it is worth refactoring this method for more explicit args and kwargs. For now, we can add a note in docstring that `non_blocking` should be passed as a keyword argument.

**Comment by daniel-bogdoll** [comment]
@qubvel Done! Thanks for the super-fast replies, was a pleasure!

File: src/transformers/tokenization_utils_base.py [hunk 1]
```diff
@@ -798,12 +798,13 @@ def as_tensor(value, dtype=None):
 
         return self
 
-    def to(self, device: Union[str, "torch.device"]) -> "BatchEncoding":
+    def to(self, device: Union[str, "torch.device"], non_blocking: bool = False) -> "BatchEncoding":
```
Comment (qubvel) [inline src/transformers/tokenization_utils_base.py:801]: To force passing `non_blocking` as a keyword argument, so later we can introduce additional positional or keyword args without breaking BC, what do you think?

```suggestion
    def to(self, device: Union[str, "torch.device"], *, non_blocking: bool = False) -> "BatchEncoding":
```

**Review by qubvel — APPROVED** [review]
Thanks for updates! Looks great, just a small suggestion

**Comment by qubvel** [comment]
@ArthurZucker or @LysandreJik please review when you have bandwidth

**Review by ArthurZucker — COMMENTED** [review]
Yeah sound super good!

File: src/transformers/tokenization_utils_base.py [hunk 2]
```diff
@@ -799,12 +799,13 @@ def as_tensor(value, dtype=None):
 
         return self
 
-    def to(self, device: Union[str, "torch.device"]) -> "BatchEncoding":
+    def to(self, device: Union[str, "torch.device"], *, non_blocking: bool = False) -> "BatchEncoding":
```
[the remark this answers is not in the document]
Comment (daniel-bogdoll) [reply src/transformers/tokenization_utils_base.py:802]: @qubvel suggested this to enforce it as a keyword argument for backwards compability

[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/tokenization_utils_base.py:802]: Thanks for explaining, good decision @qubvel ! 🤗

**Review by ArthurZucker — APPROVED** [review]

**Pull request merged**